presto-jdbc/pom.xml

```
@@ -268,6 +268,18 @@
                                     <pattern>com.fasterxml.jackson</pattern>
                                     <shadedPattern>${shadeBase}.jackson</shadedPattern>
                                 </relocation>
+                                <relocation>
+                                    <pattern>META-INF/versions/11/com.fasterxml.jackson</pattern>
+                                    <shadedPattern>META-INF/versions/11/${shadeBase}.com.fasterxml.jackson</shadedPattern>
+                                </relocation>
+                                <relocation>
+                                    <pattern>META-INF/versions/17/com.fasterxml.jackson</pattern>
+                                    <shadedPattern>META-INF/versions/17/${shadeBase}.com.fasterxml.jackson</shadedPattern>
+                                </relocation>
+                                <relocation>
+                                    <pattern>META-INF/versions/21/com.fasterxml.jackson</pattern>
+                                    <shadedPattern>META-INF/versions/21/${shadeBase}.com.fasterxml.jackson</shadedPattern>
```

Wouldn't this need to be consistent with the shaded package structure specified on line 269 (above)?

```suggestion
                                    <shadedPattern>META-INF/versions/11/${shadeBase}.jackson</shadedPattern>
                                </relocation>
                                <relocation>
                                    <pattern>META-INF/versions/17/com.fasterxml.jackson</pattern>
                                    <shadedPattern>META-INF/versions/17/${shadeBase}.jackson</shadedPattern>
                                </relocation>
                                <relocation>
                                    <pattern>META-INF/versions/21/com.fasterxml.jackson</pattern>
                                    <shadedPattern>META-INF/versions/21/${shadeBase}.jackson</shadedPattern>
```

---

The `META-INF/versions/<version>/` is prefix defined by the Multi-Release JAR spec: https://openjdk.org/jeps/238

It allows library developers to supply version of classes for newer version of Java.

For example consider you have a class `a.b.c.Service`. You want to support Java 8, but would like to leverage an API only available in Java 17 when available.

What you can do you can

- Provide a standard class supported by Java 8 and place it in default location: `a.b.c.Service`
- Provide a class supported by Java 17 and above and place it under `META-INF/versions/17/a.b.c.Service`

Now let's assume a shade plugin decides to relocate `a.b.c.Service` -> `some.other.base.a.b.c.Service`.
For the VM to find version specific classes they also need to be relocated: `META-INF/versions/17/a.b.c.Service` -> `META-INF/versions/17/some.other.base.a.b.c.Service`

---

@arhimondr understood.  But I see above we are shading to package `${shadeBase}.jackson`, not `${shadeBase}.com.fasterxml.jackson`:

```xml
                                <relocation>
                                    <pattern>com.fasterxml.jackson</pattern>
                                    <shadedPattern>${shadeBase}.jackson</shadedPattern>
                                </relocation>
```

Wouldn't this mean all multi-release classes also need to be relocated to `${shadeBase}.jackson`, not `${shadeBase}.com.fasterxml.jackson` as they are in the above code?

---

Great catch. Thank you @tdcmeehan 

Updated

Here is how it looks now:

```
$ jar -tf ./presto-jdbc/target/presto-jdbc-0.291-SNAPSHOT.jar | grep -P '(FastDoubleSwar|FastIntegerMath|BigSignificand)'
com/facebook/presto/jdbc/internal/jackson/core/io/doubleparser/BigSignificand.class
com/facebook/presto/jdbc/internal/jackson/core/io/doubleparser/FastDoubleSwar.class
com/facebook/presto/jdbc/internal/jackson/core/io/doubleparser/FastIntegerMath.class
META-INF/versions/11/com/facebook/presto/jdbc/internal/jackson/core/io/doubleparser/BigSignificand.class
META-INF/versions/11/com/facebook/presto/jdbc/internal/jackson/core/io/doubleparser/FastDoubleSwar.class
META-INF/versions/11/com/facebook/presto/jdbc/internal/jackson/core/io/doubleparser/FastIntegerMath.class
META-INF/versions/17/com/facebook/presto/jdbc/internal/jackson/core/io/doubleparser/FastDoubleSwar.class
META-INF/versions/17/com/facebook/presto/jdbc/internal/jackson/core/io/doubleparser/FastIntegerMath.class
META-INF/versions/21/com/facebook/presto/jdbc/internal/jackson/core/io/doubleparser/FastDoubleSwar.class
META-INF/versions/21/com/facebook/presto/jdbc/internal/jackson/core/io/doubleparser/FastIntegerMath.class
```